### ISSUE TITLE
convert2bed-megarow has persistent "Segmentation fault" error

### ISSUE THREAD

**Issue by SchwarzEM** [bedops/bedops#207]
I'm trying to use convert2bed-megarow to convert SAM files into BED files.  My test instance is a *small* SAM file that, in my hands, can be easily and reliably converted to BED by either of the two sibling programs of convert2bed-megarow, convert2bed-typical or convert2bed-float128:

    cat vc2010_Mt.minimap2.n2_Mt.sam | convert2bed-typical --input=sam --output=bed - 1>vc2010_Mt.minimap2.n2_Mt.typical.bed 2>vc2010_Mt.minimap2.n2_Mt.typical.err ;

    cat vc2010_Mt.minimap2.n2_Mt.sam | convert2bed-float128 --input=sam --output=bed - 1>vc2010_Mt.minimap2.n2_Mt.float128.bed 2>vc2010_Mt.minimap2.n2_Mt.float128.err ;

Both of these line commands work fine, whether I am using convert2bed from precompiled binaries for bedops 2.4.32, or from binaries that I have compiled from bedops 2.4.32 source code on my various Linux systems.

But, on three different Linux systems, when I run this:

    cat vc2010_Mt.minimap2.n2_Mt.sam | convert2bed-megarow --input=sam --output=bed - 1>vc2010_Mt.minimap2.n2_Mt.megarow.bed 2>vc2010_Mt.minimap2.n2_Mt.megarow.err ;

I get some version of the following error message:

    Segmentation fault

Note that the SAM file which I am feeding into convert2bed-megarow is small (19 KB!) and is easily handled by either convert2bed-typical or convert2bed-float128, which were both compiled simultaneously and in parallel with convert2bed-megarow.  That isolates the problem to convert2bed-megarow specifically.  I have observed this problem on multiple computers and whether I use pre-compiled or source-compiled convert2bed-megarow.  Although almost all of my tests have been with bedops 2.4.32, on one system I have also replicated these negative results with both pre-compiled and source-compiled convert2bed-megarow from the immediately preceding version (bedops 2.4.31).

There is one exception to this pattern, that is both informative and frustrating: one of my computer systems consists of a login node (on which I'm forbidden to run serious computations) and a compute node (on which serious computations are allowed).  When I run convert2bed-megarow on the *login* node, it runs flawlessly rather than giving me a segmentation error -- whether it is a precompiled binary or one that I compiled from source on the *compute* node.  Yet it fails on the compute node!

This last strange discrepancy leads me to believe that I'm getting these segmentation faults because of some subtle difference in the kernels or core operating systems of the login versus the compute nodes.  Here are the results of "uname -a" on the login node:

    Linux br005.pvt.bridges.psc.edu 3.10.0-693.21.1.el7.x86_64 #1 SMP Wed Mar 7 19:03:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

versus the compute node:

    Linux r001.pvt.bridges.psc.edu 3.10.0-693.11.6.el7.x86_64 #1 SMP Thu Jan 4 01:06:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

Note that the login node kernel (3.10.0-693.21.1.el7.x86_64) is very slightly more advanced than the compute node kernel (3.10.0-693.11.6.el7.x86_64).

To sum up: I can't use convert2bed-megarow because it seems to be almost completely unusable except in a particular system where I happen to be banned from using it in production mode.  The unusability must be due to a specific problem in the code of convert2bed-megarow alone, not convert2bed programs generally; it is evoked by *almost* all Linux environments in which I've worked, but not every single one; it therefore appears to be a toxic interaction with core Linux system components (the kernel or its associated programs).  I do not seem to be able to solve the problem by compiling from source within a given Linux system.

I would be extremely appreciative if this problem could be solved and a patch for convert2bed-megarow that made it widely usable were to be placed in the latest github version of the bedops software!

**Comment by alexpreynolds** [comment]
This is a great bug report, thanks. I'll investigate and patch once I know the cause of the problem.

**Comment by alexpreynolds** [comment]
Thanks for the excellent bug report.

The problem is unlikely to do with your kernel version. Rather, "megarow" builds of BEDOPS use constants that define a longer BED line length. The megarow version of the `convert2bed` application is running into a situation where arrays that use these constants use too much of the kernel's stack and cause a segmentation fault.

I'm "cleaning up" the last of these arrays by replacing them with heap- or dynamically-allocated pointers. Once that's done and things are ready, I'll publish a new version.

**Comment by alexpreynolds** [comment]
Pushed commit 39e29b72 to further limit use of arrays in `convert2bed`

**Comment by alexpreynolds** [comment]
v2.4.33 is out with the described patch. Please feel free to follow up if this does not resolve the issue for you. https://github.com/bedops/bedops/releases/tag/v2.4.33

**Comment by SchwarzEM** [comment]
v2.4.33 downloads cleanly via bioconda, and does indeed solve the problem that I described.
Thank you very much for your highly rapid response!

(Alas, I am now discovering a new problem ... but that's a separate issue.)
